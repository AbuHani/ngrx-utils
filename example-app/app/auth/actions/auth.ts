import { Action } from '@ngrx/store';
import { User, Authenticate } from '../models/user';

export class Login implements Action {
  readonly type = '[Auth] Login';

  constructor(public payload: Authenticate) {}
}

export class LoginSuccess implements Action {
  readonly type = '[Auth] Login Success';

  constructor(public payload: { user: User }) {}
}

export class LoginFailure implements Action {
  readonly type = '[Auth] Login Failure';

  constructor(public payload: any) {}
}

export class LoginRedirect implements Action {
  readonly type = '[Auth] Login Redirect';
}

export class Logout implements Action {
  readonly type = '[Auth] Logout';
}

export type AuthActions = Login | LoginSuccess | LoginFailure | LoginRedirect | Logout;
