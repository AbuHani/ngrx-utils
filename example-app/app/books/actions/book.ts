import { Action } from '@ngrx/store';
import { Book } from '../models/book';

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class Search implements Action {
  readonly type = '[Book] Search';

  constructor(public payload: string) {}
}

export class SearchComplete implements Action {
  readonly type = '[Book] Search Complete';

  constructor(public payload: Book[]) {}
}

export class SearchError implements Action {
  readonly type = '[Book] Search Error';

  constructor(public payload: string) {}
}

export class Load implements Action {
  readonly type = '[Book] Load';

  constructor(public payload: Book) {}
}

export class Select implements Action {
  readonly type = '[Book] Select';

  constructor(public payload: string) {}
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type BookActions = Search | SearchComplete | SearchError | Load | Select;
