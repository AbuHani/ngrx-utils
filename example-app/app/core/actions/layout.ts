import { Action } from '@ngrx/store';

export class OpenSidenav implements Action {
  readonly type = '[Layout] Open Sidenav';
}

export class CloseSidenav implements Action {
  readonly type = '[Layout] Close Sidenav';
}

export type LayoutActions = OpenSidenav | CloseSidenav;
