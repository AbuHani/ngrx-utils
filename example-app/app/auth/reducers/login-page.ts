import { AuthActions } from './../actions/auth';

export interface State {
  error: string | null;
  pending: boolean;
}

export const initialState: State = {
  error: null,
  pending: false
};

export function reducer(state = initialState, action: AuthActions): State {
  switch (action.type) {
    case '[Auth] Login':
      return {
        ...state,
        error: null,
        pending: true
      };
    case '[Auth] Login Success':
      return {
        ...state,
        error: null,
        pending: false
      };
    case '[Auth] Login Failure':
      return {
        ...state,
        error: action.payload,
        pending: false
      };
    default: {
      return state;
    }
  }
}

export const getError = (state: State) => state.error;
export const getPending = (state: State) => state.pending;
