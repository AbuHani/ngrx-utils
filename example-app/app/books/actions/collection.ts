import { Action } from '@ngrx/store';
import { Book } from '../models/book';

/**
 * Add Book to Collection Actions
 */
export class AddBook implements Action {
  readonly type = '[Collection] Add Book';

  constructor(public payload: Book) {}
}

export class AddBookSuccess implements Action {
  readonly type = '[Collection] Add Book Success';

  constructor(public payload: Book) {}
}

export class AddBookFail implements Action {
  readonly type = '[Collection] Add Book Fail';

  constructor(public payload: Book) {}
}

/**
 * Remove Book from Collection Actions
 */
export class RemoveBook implements Action {
  readonly type = '[Collection] Remove Book';

  constructor(public payload: Book) {}
}

export class RemoveBookSuccess implements Action {
  readonly type = '[Collection] Remove Book Success';

  constructor(public payload: Book) {}
}

export class RemoveBookFail implements Action {
  readonly type = '[Collection] Remove Book Fail';

  constructor(public payload: Book) {}
}

/**
 * Load Collection Actions
 */
export class Load implements Action {
  readonly type = '[Collection] Load';
}

export class LoadSuccess implements Action {
  readonly type = '[Collection] Load Success';

  constructor(public payload: Book[]) {}
}

export class LoadFail implements Action {
  readonly type = '[Collection] Load Fail';

  constructor(public payload: any) {}
}

export type CollectionActions =
  | AddBook
  | AddBookSuccess
  | AddBookFail
  | RemoveBook
  | RemoveBookSuccess
  | RemoveBookFail
  | Load
  | LoadSuccess
  | LoadFail;
