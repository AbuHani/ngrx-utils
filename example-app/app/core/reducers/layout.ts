import { LayoutActions } from '../actions/layout';

export interface State {
  showSidenav: boolean;
}

const initialState: State = {
  showSidenav: false
};

export function reducer(state: State = initialState, action: LayoutActions): State {
  switch (action.type) {
    case '[Layout] Close Sidenav':
      return {
        showSidenav: false
      };

    case '[Layout] Open Sidenav':
      return {
        showSidenav: true
      };

    default:
      return state;
  }
}

export const getShowSidenav = (state: State) => state.showSidenav;
