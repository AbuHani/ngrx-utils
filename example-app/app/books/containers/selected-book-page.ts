import { Component, ChangeDetectionStrategy } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { Select } from '@ngrx-utils/store';

import * as fromBooks from '../reducers';
import * as collection from '../actions/collection';
import { Book } from '../models/book';

@Component({
  selector: 'bc-selected-book-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <bc-book-detail
      [book]="book$ | async"
      [inCollection]="isSelectedBookInCollection$ | async"
      (add)="addToCollection($event)"
      (remove)="removeFromCollection($event)">
    </bc-book-detail>
  `
})
export class SelectedBookPageComponent {
  @Select(fromBooks.getSelectedBook) book$: Observable<Book>;
  @Select(fromBooks.isSelectedBookInCollection) isSelectedBookInCollection$: Observable<boolean>;

  constructor(private store: Store<fromBooks.State>) {}

  addToCollection(book: Book) {
    this.store.dispatch(new collection.AddBook(book));
  }

  removeFromCollection(book: Book) {
    this.store.dispatch(new collection.RemoveBook(book));
  }
}
