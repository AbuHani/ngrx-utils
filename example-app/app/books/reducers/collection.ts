import { CollectionActions } from './../actions/collection';

export interface State {
  loaded: boolean;
  loading: boolean;
  ids: string[];
}

const initialState: State = {
  loaded: false,
  loading: false,
  ids: []
};

export function reducer(state = initialState, action: CollectionActions): State {
  switch (action.type) {
    case '[Collection] Load': {
      return {
        ...state,
        loading: true
      };
    }

    case '[Collection] Load Success': {
      return {
        loaded: true,
        loading: false,
        ids: action.payload.map(book => book.id)
      };
    }

    case '[Collection] Add Book Success':
    case '[Collection] Remove Book Fail': {
      if (state.ids.indexOf(action.payload.id) > -1) {
        return state;
      }

      return {
        ...state,
        ids: [...state.ids, action.payload.id]
      };
    }

    case '[Collection] Add Book Fail':
    case '[Collection] Remove Book Success': {
      return {
        ...state,
        ids: state.ids.filter(id => id !== action.payload.id)
      };
    }

    default: {
      return state;
    }
  }
}

export const getLoaded = (state: State) => state.loaded;

export const getLoading = (state: State) => state.loading;

export const getIds = (state: State) => state.ids;
