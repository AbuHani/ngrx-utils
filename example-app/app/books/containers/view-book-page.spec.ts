import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { ActivatedRoute } from '@angular/router';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { MatCardModule } from '@angular/material';
import { NgrxSelect, NgrxUtilsModule } from '@ngrx-utils/store';

import { ViewBookPageComponent } from './view-book-page';
import * as book from '../actions/book';
import * as fromBooks from '../reducers';
import { SelectedBookPageComponent } from './selected-book-page';
import { BookDetailComponent } from '../components/book-detail';
import { BookAuthorsComponent } from '../components/book-authors';
import { AddCommasPipe } from '../../shared/pipes/add-commas';

describe('View Book Page', () => {
  let params = new BehaviorSubject({});
  let fixture: ComponentFixture<ViewBookPageComponent>;
  let store: Store<fromBooks.State>;
  let instance: ViewBookPageComponent;
  let ngrxSelect: NgrxSelect;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [MatCardModule, NgrxUtilsModule],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: { params }
        },
        {
          provide: Store,
          useValue: {
            select: jest.fn(),
            next: jest.fn(),
            pipe: jest.fn()
          }
        }
      ],
      declarations: [
        ViewBookPageComponent,
        SelectedBookPageComponent,
        BookDetailComponent,
        BookAuthorsComponent,
        AddCommasPipe
      ]
    });

    fixture = TestBed.createComponent(ViewBookPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);
    ngrxSelect = TestBed.get(NgrxSelect);
    ngrxSelect.connect(store);
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a book.Select action on init', () => {
    const action = new book.Select('2');
    params.next({ id: '2' });

    fixture.detectChanges();

    expect(store.next).toHaveBeenLastCalledWith(action);
  });
});
