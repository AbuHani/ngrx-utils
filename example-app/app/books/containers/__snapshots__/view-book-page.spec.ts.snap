// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Book Page should compile 1`] = `
<bc-view-book-page
  ngOnDestroy={[Function Function]}
>
  
      <bc-selected-book-page>
    
      
    <bc-book-detail
      _nghost-c0=""
    >
      
      
      
      
  
    
    </bc-book-detail>
    
    
  </bc-selected-book-page>
    
</bc-view-book-page>
`;
